src/allmydata/test/test_runner.py

```diff
@@ -621,3 +627,57 @@ def _bad_directory_test(self, workdir, description, operation, expected_message)
         # What's left is a perfect indicator that the process has exited and
         # we won't get blamed for leaving the reactor dirty.
         yield client_running
+
+
+class OnStdinCloseTests(SyncTestCase):
+    """
+    Tests for on_stdin_close
+    """
+
+    def test_close_called(self):
+        """
+        our on-close method is called when stdin closes
+        """
+        reactor = MemoryReactorClock()
+        called = []
+
+        def onclose():
+            called.append(True)
+        proto = on_stdin_close(reactor, onclose)
+        self.assertEqual(called, [])
+
+        # on Unix we can just close all the readers, correctly
+        # "simulating" a stdin close .. of course, Windows has to be
+        # difficult
+        if platform.isWindows():
+            proto.writeConnectionLost()
+            proto.readConnectionLost()
```

Comment:
This was pretty confusing until I recognized that `proto` is a `StandardIO` and `writeConnectionLost` / `readConnectionLost` here don't refer to `IHalfCloseableProtocol` but just happen to be the names of the Windows `StandardIO` implementation details for stdin/stdout closed notification. :/

Too bad Twisted doesn't provide a usable `StandardIO` test double, I guess?  Maybe we can ask for one.

Since lines 649-658 are duplicated in the next test, I'd at least pull them out into a shared helper function.